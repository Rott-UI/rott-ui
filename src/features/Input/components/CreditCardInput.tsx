// React Imports
import {FC} from 'react'

// React Native Imports
import {StyleSheet} from 'react-native'

// Component Imports
import {InputProps} from '../models'
import {InputStyles} from '../styles'

// Package Imports
import {MaskedTextInput} from 'react-native-mask-text'

export const CreditCardInput: FC<Omit<InputProps, 'label' | 'name'>> = ({
  fontSize,
  onChangeText,
  theme,
  disabled,
  size,
  ...props
}) => {
  const handleOnChangeText = (text: string) => {
    const replacedText = text.replace(/[^0-9]/g, '')
    onChangeText!(replacedText)
  }

  return (
    <MaskedTextInput
      editable={!disabled}
      mask='9999-9999-9999-9999'
      placeholder='**** **** **** ****'
      keyboardType='number-pad'
      maxLength={19}
      onChangeText={(text, rawText) => handleOnChangeText(rawText)}
      style={StyleSheet.flatten([InputStyles({fontSize, theme, size}).defaultTextInputStyle])}
      {...props}
    />
  )
}
