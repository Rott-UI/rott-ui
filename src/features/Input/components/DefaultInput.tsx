// React Imports
import {FC} from 'react'

// React Native Imports
import {StyleSheet, TextInput} from 'react-native'

// Component Imports
import {InputProps} from '../models'
import {InputStyles} from '../styles'

export const DefaultInput: FC<Omit<InputProps, 'name'>> = ({
  label,
  placeholder,
  fontSize,
  theme,
  size,
  ...props
}) => {
  return (
    <TextInput
      keyboardType='default'
      autoCapitalize='none'
      placeholder={placeholder ?? (typeof label === 'string' ? label : undefined)}
      style={StyleSheet.flatten([InputStyles({fontSize, theme, size}).defaultTextInputStyle])}
      {...props}
    />
  )
}
