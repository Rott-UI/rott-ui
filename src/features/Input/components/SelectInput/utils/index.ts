export * from './listHeightNormalizer'
export * from './modalHeightPercentageNormalizer'
