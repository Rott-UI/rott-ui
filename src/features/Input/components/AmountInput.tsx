// React Imports
import {FC, useCallback, useEffect, useRef, useState} from 'react'

// React Native Imports
import {StyleSheet, TextInput} from 'react-native'

// Component Imports
import {InputProps} from '../models'
import {InputStyles} from '../styles'
import {Item} from '../../Item'
import {Label} from '../../Label'
import {Icon} from '../../Icon'

// Util and Lib Imports
import {themeConfig} from '../../../providers'

// Package Imports
import {AmountInputStyles} from '../styles/AmountInput.style'
import {InputStyleNormalizer} from '../utils/inputNormalizer'

export const AmountInput: FC<Omit<InputProps, 'placeholder' | 'label' | 'name'>> = ({
  fontSize,
  onChangeText,
  value,
  theme,
  disabled,
  size,
  ...props
}) => {
  const amountRef = useRef<TextInput>(null)
  const [amount, setAmount] = useState('')
  const [currency, setCurrency] = useState('')

  const resetInternalStates = useCallback(() => {
    setAmount('')
    setCurrency('')
    onChangeText!('0.00')
  }, [])

  const replaceTextWithNumberOrEmpty = (text: string) =>
    text.replace(/[^0-9]/g, '').replace(/^0+/, '')

  const amountNormalizer = (text: string) => {
    if (!text) return resetInternalStates()

    const replacedAmount = replaceTextWithNumberOrEmpty(text)
    const reversedText = replacedAmount.split('').reverse().join('') ?? ''
    const dottedText = reversedText?.match(/.{1,3}/g)?.join('.')
    const normalizedText = dottedText?.split('').reverse().join('')

    setAmount(normalizedText ?? '')
  }

  const handleTextChange = (inputTextAmount: string, inputTextCurrency: string) => {
    if (!inputTextAmount && !inputTextCurrency) return

    let amountFormat = replaceTextWithNumberOrEmpty(inputTextAmount)
    let currencyFormat = replaceTextWithNumberOrEmpty(inputTextCurrency)

    amountFormat = amountFormat.length === 0 ? '0' : amountFormat

    currencyFormat = currencyFormat.length === 0 ? '00' : currencyFormat
    currencyFormat = currencyFormat.length === 1 ? `${currencyFormat}0` : currencyFormat

    onChangeText!(`${amountFormat}.${currencyFormat}`)
  }

  const placeholderColorNormalizer =
    theme === 'dark'
      ? themeConfig.colors['white']
      : (!amount && !currency) ||
        ((amount === '0' || amount === '') && (!currency || currency === '' || currency === '00'))
      ? themeConfig.colors['grey-200']
      : themeConfig.colors['grey-900']

  useEffect(() => {
    const initialAmount = value?.split(',')[0]
    const initialCurrency = value?.split(',')[1]

    setAmount(initialAmount ?? '')
    setCurrency(initialCurrency ?? '')
  }, [])

  useEffect(() => {
    if (value === '' || !value || value === '0') resetInternalStates()
  }, [value])

  useEffect(() => {
    handleTextChange(amount, currency)
  }, [amount, currency])

  return (
    <Item relative onTouchStart={() => amountRef.current?.focus()}>
      <Item row {...props}>
        <TextInput
          ref={amountRef}
          nativeID='amount-native-id'
          testID='amount-test-id'
          editable={!disabled}
          placeholder='0'
          maxLength={11}
          value={amount}
          keyboardType='number-pad'
          onChangeText={(text) => amountNormalizer(text)}
          style={StyleSheet.flatten([
            InputStyles({fontSize, theme, size}).defaultTextInputStyle,
            AmountInputStyles().amountInputStyle,
            {
              color: placeholderColorNormalizer,
            },
          ])}
          placeholderTextColor={placeholderColorNormalizer}
        />

        <Item
          justifyContentFlexEnd
          paddingBottom={InputStyleNormalizer({size}).bottomElementPadding}>
          <Label
            fontSize={
              fontSize ??
              InputStyleNormalizer({
                size,
              }).placeholderSize
            }
            fontWeight='bold'
            color={placeholderColorNormalizer as string}>
            ,
          </Label>
        </Item>

        <TextInput
          testID='currency-test-id'
          maxLength={2}
          placeholder='00'
          value={currency}
          keyboardType='number-pad'
          onChangeText={(text) => setCurrency(replaceTextWithNumberOrEmpty(text).substring(0, 2))}
          onFocus={() => (currency === '00' ? setCurrency('') : undefined)}
          onBlur={() => (currency.isEmpty() || currency === '0' ? setCurrency('00') : undefined)}
          style={StyleSheet.flatten([
            InputStyles({fontSize, theme, size}).defaultTextInputStyle,
            AmountInputStyles().amountInputStyle,
            {
              color: placeholderColorNormalizer,
            },
          ])}
          placeholderTextColor={placeholderColorNormalizer}
        />
      </Item>

      <Item absolute right={0} bottom={InputStyleNormalizer({size}).icon.paddingBottom}>
        <Icon
          name='TL'
          width={InputStyleNormalizer({size}).icon.width}
          height={InputStyleNormalizer({size}).icon.height}
          color={themeConfig.colors['grey-200']}
          mode='fill'
          noStroke
        />
      </Item>
    </Item>
  )
}
