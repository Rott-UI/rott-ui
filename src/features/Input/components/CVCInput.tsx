// React Imports
import {FC} from 'react'

// React Native Imports
import {Alert, StyleSheet, TextInput} from 'react-native'

// Component Imports
import {InputProps} from '../models'
import {InputStyles, CVCInputStyles} from '../styles'
import {Item} from '../../Item'
import {Icon} from '../../Icon'
import {Pressable} from '../../Pressable'

// Util and Lib Imports
import {formatMessage} from '../../../utils'

export const CVCInput: FC<Omit<InputProps, 'label' | 'placeholder' | 'name'>> = ({
  fontSize,
  onChangeText,
  theme,
  size,
  ...props
}) => {
  const handleTextChange = (inputText: string) => {
    let text = inputText.replace(/[^0-9]/g, '')
    text = text.length > 3 ? text.substring(0, 3) : text
    onChangeText!(text)
  }

  return (
    <Item row>
      <TextInput
        editable={!props.disabled}
        placeholder='***'
        style={StyleSheet.flatten([InputStyles({fontSize, theme, size}).defaultTextInputStyle])}
        keyboardType='number-pad'
        maxLength={3}
        onChangeText={handleTextChange}
        {...props}
      />

      {/* TODO: CVC Info butonuna kart arkayüzü eklenmeli */}
      <Pressable
        testID='info-icon-test-id'
        style={CVCInputStyles().infoIcon}
        justifyContentCenter
        alignItemsCenter
        onPress={() =>
          Alert.alert(formatMessage('CVC.INFO.TITLE'), formatMessage('CVC.INFO.DESCRIPTION'), [
            {
              text: formatMessage('COMMON.OK'),
              style: 'cancel',
            },
          ])
        }>
        <Icon name='CHECK_CIRCLE' height={24} width={24} variant='grey-900' />
      </Pressable>
    </Item>
  )
}
