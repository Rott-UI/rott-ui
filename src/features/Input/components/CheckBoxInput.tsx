// React Imports
import {FC, ReactNode} from 'react'

// React Native Imports

// Component Imports
import {InputProps} from '../models'
import {Label} from '../../Label'
import {Item} from '../../Item'
import {Pressable} from '../../Pressable'

// Util and Lib Imports
import {InputStyles} from '../styles'
import {InputStyleNormalizer} from '../utils/inputNormalizer'
import {themeConfig} from '../../../providers'

interface CheckBoxProps extends Omit<InputProps, 'placeholder' | 'name' | 'label'> {
  description?: ReactNode
  checked?: boolean
  onCheckChange?: (checked: boolean) => void
}

export const CheckBoxInput: FC<CheckBoxProps> = ({
  description,
  checked,
  fontSize,
  theme,
  size,
  disabled,
  onCheckChange,
}) => {
  return (
    <Pressable
      flex={0}
      testID='checkbox-input-test-id'
      onPress={() => !disabled && onCheckChange!(!checked)}>
      <Item
        row
        alignItemsCenter
        style={InputStyles({theme, size}).defaultTextInputStyle}
        testID='checkbox-container-test-id'>
        <Item
          borderRadius={8}
          borderColor={checked ? themeConfig.colors['primary'] : themeConfig.colors['grey-200']}
          borderWidth={2}
          backgroundColor='white'
          width={24}
          height={24}
          alignItemsCenter
          justifyContentCenter>
          {checked && (
            <Item
              testID='checkbox-checked-test-id'
              borderWidth={2}
              borderColor={themeConfig.colors['neutral-blue-alpha']}
              borderRadius={4}
              backgroundColor={themeConfig.colors['primary']}
              width={15}
              height={15}
            />
          )}
        </Item>

        {description && typeof description === 'string' && (
          <Label
            testID='checkbox-default-label-test-id'
            fontSize={fontSize ?? InputStyleNormalizer({size}).placeholderSize}
            fontWeight='bold'
            marginLeft={12}
            variant={theme === 'dark' ? 'white' : 'grey-900'}>
            {description}
          </Label>
        )}

        {description && typeof description !== 'string' && <>{description}</>}
      </Item>
    </Pressable>
  )
}
