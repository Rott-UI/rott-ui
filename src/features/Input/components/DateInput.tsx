// React Native Imports
// import {Modal} from 'react-native'

// React Imports
import React, {FC, useState} from 'react'

// Component Imports
import {InputProps} from '../models'
import {Item} from '../../Item'
import {Pressable} from '../../Pressable'
import {DateInputStyles} from '../styles'
import {Modal, useModal} from '../../Modal'

// Feature Imports
import {getLanguageState} from '../../../features/app'

// Util and Lib Imports
import {display} from '../../../utils'
import {formatByDateMode} from '../utils'

// Package Imports
import DatePicker from 'react-native-date-picker'
import {COLOURS} from '../../../constants'
import {InputStyleNormalizer} from '../utils/inputNormalizer'
import {useAppSelector} from '../../../hooks'

interface DateInputProps extends Omit<InputProps, 'label' | 'name'> {}

/**
 *
 * DateInput'un kullanımı diğer inputlara göre biraz daha farklı.
 * Yapısı gereği diğer inputlarda kullandığımız "onChangeText" event'i ile değişen değeri almamız mümkün olmadığı için
 * formik'in içinde yer alan "setFieldValue" metodunu kullanmamız gerek.
 *
 * ### **Örnek**:
 * <Input onDateChange={(date) => setFieldValue('date', date)} />
 *
 * @property {(date: Date) => void)} onDateChange: seçilen tarih değerini almak içim bu property'nin kullanılması gerekmektedir.
 * @property {'date' | 'time' | 'datetime'} mode: İstenilen mod neyse ona göre belirtilmelidir. Default değer datetime olarak ayarlanmıştır.
 * @property {boolean} dontAllowClearing: Eğer date input'un değerinin temizlenmesine izin verilmemesi gerekiyorsa bu propert kullanılmalıdır.
 * dontAllowClearing={true} yerine sadece downAllowClearing olarak verirseniz değer otomatik olarak true kabul edilir. False olmadığı durumlarda
 * lütfen ={true} şeklinde tanımlama yapmayınız.
 */
export const DateInput: FC<DateInputProps> = ({
  placeholder = 'Lütfen bir tarih seçiniz',
  onDateChange,
  mode = 'date',
  allowClear = false,
  minimumDate,
  maximumDate,
  value,
  testID,
  theme,
  size,
  ...props
}) => {
  const appLanguage = useAppSelector(getLanguageState)

  const [datePickerValue, setDatePickerValue] = useState<Date>(new Date())

  const handleConfirmPress = () => {
    let validDate: Date
    if (minimumDate && datePickerValue.getTime() < minimumDate?.getTime()) validDate = minimumDate
    else if (maximumDate && datePickerValue.getTime() > maximumDate?.getTime())
      validDate = maximumDate
    else validDate = datePickerValue

    onDateChange!(validDate)
    Modal.hideModal(98)
  }

  const handleClearPress = () => {
    hideModal()
    setDatePickerValue(new Date())
    onDateChange!(null as any)
  }

  const {showModal, hideModal} = useModal(
    {
      id: 98,
      testID: 'date-input-modal',
      visible: true,
      height: 30,
      header: (
        <Item
          justifyContentCenter
          backgroundColor={COLOURS.GREY900}
          height={display.heightPixel(50)}
          style={DateInputStyles().dateInputHeaderStyle}>
          <Pressable
            testID='date-input-confirm-button'
            onPress={() => handleConfirmPress()}
            text='Tamam'
            textVariant='white'
            textSize='xl'
            textWeight={700}
            style={DateInputStyles().confirmButtonStyle}
          />

          {allowClear && value && (
            <Pressable
              testID='date-input-clear-button'
              onPress={handleClearPress}
              text='Temizle'
              textVariant='white'
              textSize='xl'
              textWeight={700}
              style={DateInputStyles().cancelButtonStyle}
            />
          )}
        </Item>
      ),
      children: (
        <Item size='full' alignItemsCenter justifyContentCenter>
          <DatePicker
            testID={testID}
            mode={mode}
            date={datePickerValue}
            textColor={COLOURS.GREY900}
            locale={appLanguage?.name}
            minimumDate={minimumDate}
            maximumDate={maximumDate}
            onDateChange={(date) => setDatePickerValue(date)}
            {...props}
          />
        </Item>
      ),
    },
    [datePickerValue]
  )

  return (
    <Item row>
      <Pressable
        size='full'
        height={InputStyleNormalizer({size}).height}
        testID='date-input-value-container'
        flex={0}
        justifyContentCenter
        textSize='xl'
        text={value ? formatByDateMode(mode, value) : placeholder}
        textStyle={DateInputStyles({theme, value}).pressableTextStyle}
        textVariant={theme === 'dark' ? 'white' : value ? 'grey-900' : 'grey-200'}
        onPress={() => {
          showModal()
        }}
        marginTop='2.5%'
      />
    </Item>
  )
}
