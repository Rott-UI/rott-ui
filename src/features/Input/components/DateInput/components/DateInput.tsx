/* eslint-disable indent */

// React Imports
import React, {FC, useEffect} from 'react'

// Component Imports
import {DataModel, InputProps} from '../../../models'
import {Item} from '../../../../Item'
import {Pressable} from '../../../../Pressable'
import {DateInputStyles} from '../../../styles'
import {Modal, useModal} from '../../../../Modal'
import {Icon} from '../../../../Icon'
import {List} from '../../../../List'
import {formatByDateMode} from '../../../utils'
import {InputStyleNormalizer} from '../../../utils/inputNormalizer'
import {ModalIdEnum} from '../../../../models'
import {Label} from '../../../../Label'
import {Separator} from '../../../../Separator'

// Util and Lib Imports
import {display, formatMessage} from '../../../../../utils'

// Package Imports
import DatePicker from 'react-native-date-picker'
import {themeConfig} from '../../../../../providers'

interface DateInputProps extends Omit<InputProps, 'label' | 'name' | 'mode'> {
  mode?: 'date' | 'time' | 'datetime' | 'modal-date' | 'modal-time' | 'modal-datetime'
  data?: DataModel[]
}

/**
 *
 * DateInput'un kullanımı diğer inputlara göre biraz daha farklı.
 * Yapısı gereği diğer inputlarda kullandığımız "onChangeText" event'i ile değişen değeri almamız mümkün olmadığı için
 * formik'in içinde yer alan "setFieldValue" metodunu kullanmamız gerek.
 *
 * ### **Örnek**:
 *   <Input
      label='Date Deneme'
      name='order-date'
      onDateChange={(date) => setFieldValue('orderDate', date)}
      type='date'
      mode='modal-date'
      value={orderDate}
      data={[
        {
          label: 'Bugün',
          value: new Date(),
        },
        {
          label: 'Yarın',
          value: add(new Date(), {days: 1}),
        },
      ]}
    />
 *
 * @property {(date: Date) => void)} onDateChange: seçilen tarih değerini almak içim bu property'nin kullanılması gerekmektedir.
 * @property {'date' | 'time' | 'datetime'} mode: İstenilen mod neyse ona göre belirtilmelidir. Default değer datetime olarak ayarlanmıştır.
 * @property {boolean} dontAllowClearing: Eğer date input'un değerinin temizlenmesine izin verilmemesi gerekiyorsa bu propert kullanılmalıdır.
 * dontAllowClearing={true} yerine sadece downAllowClearing olarak verirseniz değer otomatik olarak true kabul edilir. False olmadığı durumlarda
 * lütfen ={true} şeklinde tanımlama yapmayınız.
 */
let externalDate = new Date()
export const DateInput: FC<DateInputProps> = ({
  placeholder = 'Lütfen bir tarih seçiniz',
  onDateChange,
  mode = 'date',
  allowClear = false,
  minimumDate,
  maximumDate,
  value,
  testID,
  theme,
  data,
  disabled,
  size,
  ...props
}) => {
  const appLanguage = themeConfig.options.appLanguage

  const handleConfirmPress = (date?: Date) => {
    let validDate: Date

    if (minimumDate && (date ? date.getTime() : externalDate.getTime()) < minimumDate?.getTime())
      validDate = minimumDate
    else if (
      maximumDate &&
      (date ? date.getTime() : externalDate.getTime()) > maximumDate?.getTime()
    )
      validDate = maximumDate
    else validDate = date ?? externalDate

    onDateChange!(validDate)
    Modal.hideModal(ModalIdEnum.NativeDatePicker)
  }

  const handleClearPress = () => {
    hideModal()
    externalDate = new Date()
    onDateChange!(null as any)
  }

  const isSameDay = (date1: Date, date2: Date) =>
    date1?.getDate() === date2?.getDate() &&
    date1?.getMonth() === date2?.getMonth() &&
    date1?.getFullYear() === date2?.getFullYear()

  const modalHeightCalculator = (dataCount?: number) => {
    if (dataCount === 0 || !dataCount) return 30

    const headerHeightInPercentage = 12.1 // 102 / 8.44
    const dataHeight = (dataCount + 2) * 52 // +1 for the default date picker item
    const calculatedModalHeightPercentage = dataHeight / 8.44 + headerHeightInPercentage

    return calculatedModalHeightPercentage >= 100 ? 100 : calculatedModalHeightPercentage
  }
  const {showModal: showNativeDatePicker, hideModal} = useModal(
    {
      id: ModalIdEnum.NativeDatePicker,
      testID: 'date-input-modal',
      visible: true,
      height: modalHeightCalculator(),
      header: (
        <Item
          justifyContentCenter
          backgroundColor={themeConfig.colors['grey-900']}
          height={50}
          style={DateInputStyles().dateInputHeaderStyle}>
          <Pressable
            testID='date-input-confirm-button'
            onPress={() => handleConfirmPress()}
            text='Tamam'
            textVariant='white'
            textSize='xl'
            textWeight={700}
            style={DateInputStyles().confirmButtonStyle}
          />

          {allowClear && value && (
            <Pressable
              testID='date-input-clear-button'
              onPress={handleClearPress}
              text='Temizle'
              textVariant='white'
              textSize='xl'
              textWeight={700}
              style={DateInputStyles().cancelButtonStyle}
            />
          )}
        </Item>
      ),
      children: (
        <Item size='full' alignItemsCenter justifyContentCenter>
          <DatePicker
            testID={testID}
            mode={mode?.replace('modal-', '') as any}
            date={value ? new Date(value) : new Date()}
            textColor={themeConfig.colors['grey-900']}
            locale={appLanguage?.name}
            minimumDate={minimumDate}
            maximumDate={maximumDate}
            onDateChange={(date: Date) => (externalDate = date)}
            {...props}
          />
        </Item>
      ),
    },
    []
  )
  const {showModal} = useModal()

  const showModalDatePicker = () =>
    showModal({
      id: ModalIdEnum.DatePickerModal,
      testID: 'date-input-modal-picker-test-id',
      visible: true,
      height: modalHeightCalculator(data?.length),
      slideToClose: true,
      closeButton: false,
      onClose: () => hideModal(ModalIdEnum.DatePickerModal),
      headerBackgroundColor: 'grey-900',
      panResponderBackgroundColor: 'grey-900',
      header: {
        leftIcon: {
          testID: 'cancel-button-test-id',
          name: 'CHEVRON_LEFT',
          mode: 'stroke',
          width: 40,
          height: 40,
          strokeWidth: 2,
          backgroundColor: themeConfig.colors['grey-800'],
          borderColor: themeConfig.colors['primary'],
          borderRadius: 24,
          alignItemsCenter: true,
          onPress: () => hideModal(ModalIdEnum.DatePickerModal),
        },
        title: 'İşlem Tarihi',
      },
      children: (
        <Item paddingTop={16} flex={1} backgroundColor={themeConfig.colors['grey-900']}>
          <Separator
            width='full'
            height={1}
            backgroundColor={themeConfig.colors['neutral-grey-alpha-200']}
          />
          <List
            height={display.setHeight(modalHeightCalculator(data?.length))}
            data={
              data
                ? [
                    ...data,
                    {
                      label: formatMessage('COMMON.SELECT.DATE'),
                      action: () => {
                        hideModal(ModalIdEnum.DatePickerModal)
                        showNativeDatePicker()
                      },
                    },
                  ]
                : []
            }
            renderItem={({item}) => {
              const selected = item.value
                ? isSameDay(item.value, new Date(value!))
                : item.value
                ? true
                : false

              return (
                <>
                  <Pressable
                    testID='date-input-modal-picker-item-test-id'
                    onPress={() => {
                      if (item.value) {
                        externalDate = item.value
                        handleConfirmPress(item.value)
                        Modal.hideModal()
                      } else item.action()
                    }}
                    height={52}
                    justifyContentCenter
                    paddingHorizontal={24}>
                    <Item row size='full' relative>
                      <Label variant='white' fontSize='lg' fontFamily='Markpro-Medium'>
                        {item.label}
                      </Label>

                      <Item
                        width={24}
                        height={24}
                        borderRadius={24}
                        borderWidth={2}
                        borderColor={
                          selected ? themeConfig.colors['primary'] : themeConfig.colors['grey-200']
                        }
                        justifyContentCenter
                        alignItemsCenter
                        absolute
                        right={0}>
                        {selected && (
                          <Item
                            width={16}
                            height={16}
                            borderRadius={16}
                            backgroundColor={themeConfig.colors['primary']}
                            // eslint-disable-next-line react-native/no-inline-styles
                            style={{
                              shadowColor: themeConfig.colors['neutral-blue-soft'],
                              shadowOffset: {
                                width: 0,
                                height: 0,
                              },
                              shadowOpacity: 1,
                              shadowRadius: 1,
                              borderWidth: 2,
                              borderColor: themeConfig.colors['neutral-blue-soft'],
                            }}
                          />
                        )}
                      </Item>
                    </Item>
                  </Pressable>
                  <Separator
                    width='full'
                    height={1}
                    backgroundColor={themeConfig.colors['neutral-grey-alpha-200']}
                  />
                </>
              )
            }}
          />
        </Item>
      ),
    })

  useEffect(() => {
    externalDate = new Date(value ? new Date(value) : new Date())

    return () => {
      externalDate = new Date()
    }
  }, [])

  return (
    <Item row>
      <Pressable
        size='full'
        height={InputStyleNormalizer({size}).height}
        testID='date-input-value-container'
        flex={0}
        justifyContentCenter
        textSize='xl'
        text={
          data && value
            ? data.find((item) => {
                const valueAsDate = new Date(value)

                return isSameDay(item.value!, valueAsDate)
              })?.label ??
              formatByDateMode(mode.replace('modal-', '') as 'date' | 'time' | 'datetime', value)
            : value
            ? formatByDateMode(mode, value)
            : placeholder
        }
        textStyle={DateInputStyles().pressableTextStyle}
        textVariant={theme === 'dark' ? 'white' : value ? 'grey-900' : 'grey-200'}
        onPress={() => {
          !disabled && mode.includes('modal') ? showModalDatePicker() : showNativeDatePicker()
        }}
      />

      <Item absolute right={0} bottom={InputStyleNormalizer({size}).icon.paddingBottom}>
        <Icon
          name='CALENDAR'
          width={InputStyleNormalizer({size}).icon.width}
          height={InputStyleNormalizer({size}).icon.height}
          color={themeConfig.colors['grey-200']}
          mode='stroke'
        />
      </Item>
    </Item>
  )
}
