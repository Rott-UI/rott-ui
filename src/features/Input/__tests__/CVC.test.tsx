// Component Imports
import {CVCInput} from '../components'

// Util and Lib Imports
import {render, fireEvent} from '../../../utils'

describe('CVC Input -> Custom Input', () => {
  const inputTestId = 'input-test-id'
  const cvcIconTestId = 'info-icon-test-id'

  it('ilk render anında snapshot ile eşleşmeli', () => {
    const renderedInput = render(<CVCInput testID={inputTestId} />)

    expect(renderedInput).toMatchSnapshot()
  })

  it('ilk renderlandiginda info iconu gorunmeli', () => {
    const {queryByTestId} = render(<CVCInput testID={inputTestId} />)

    const cvcInputElement = queryByTestId(cvcIconTestId)

    expect(cvcInputElement).not.toBeNull()
  })

  it('ilk renderlandiginda ilk renderlandiginda içerik boş olmalı', () => {
    const {getByTestId} = render(<CVCInput testID={inputTestId} />)

    const cvcInputElement = getByTestId(inputTestId)

    expect(cvcInputElement.props.value).toBeUndefined()
  })

  it('max karakter 3 olmali', () => {
    const onChangeTextMock = jest.fn()
    const {getByTestId} = render(<CVCInput testID={inputTestId} onChangeText={onChangeTextMock} />)
    const cvcInputElement = getByTestId(inputTestId)
    fireEvent.changeText(cvcInputElement, '125552')

    expect(onChangeTextMock).toHaveBeenCalledWith('125')
  })

  it('numerik olmalı', () => {
    const onChangeTextMock = jest.fn()
    const {getByTestId} = render(<CVCInput testID={inputTestId} onChangeText={onChangeTextMock} />)
    let cvcInputElement = getByTestId(inputTestId)
    const testText = '1A2b!'

    fireEvent.changeText(cvcInputElement, testText)

    expect(onChangeTextMock).toHaveBeenCalledWith('12')
  })

  it('cvc input render olduğu zaman klavye olarak number-pad ekranda görülmeli', () => {
    const {getByTestId} = render(<CVCInput testID={inputTestId} />)
    const cvcInputElement = getByTestId(inputTestId)

    expect(cvcInputElement.props.keyboardType).toBe('number-pad')
  })

  // TODO: bu test refactor edilecek
  // it('cvc input tıklandığı zaman ekranda klavye gözükmeli', async () => {
  //   const keyboardDidShowMock = jest.fn()
  //   Keyboard.addListener('keyboardDidShow', keyboardDidShowMock)

  //   const {getByTestId} = render(<CVCInput testID={inputTestId} />)

  //   await act(() => {
  //     fireEvent(getByTestId(inputTestId), 'press')
  //   })

  //   expect(keyboardDidShowMock).not.toHaveBeenCalled()
  //   waitFor(() => expect(keyboardDidShowMock).toHaveBeenCalled())
  // })
})
