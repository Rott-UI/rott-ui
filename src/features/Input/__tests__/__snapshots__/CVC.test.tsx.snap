// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CVC Input -> Custom Input ilk render anında snapshot ile eşleşmeli 1`] = `
<View
  style={
    {
      "alignItems": undefined,
      "backgroundColor": undefined,
      "bottom": undefined,
      "flex": undefined,
      "flexDirection": "row",
      "flexGrow": undefined,
      "gap": undefined,
      "height": undefined,
      "justifyContent": undefined,
      "left": undefined,
      "marginBottom": undefined,
      "marginLeft": undefined,
      "marginRight": undefined,
      "marginTop": undefined,
      "maxHeight": undefined,
      "maxWidth": undefined,
      "minHeight": undefined,
      "minWidth": undefined,
      "overflow": undefined,
      "paddingBottom": undefined,
      "paddingHorizontal": undefined,
      "paddingLeft": undefined,
      "paddingRight": undefined,
      "paddingTop": undefined,
      "paddingVertical": undefined,
      "position": undefined,
      "right": undefined,
      "top": undefined,
      "width": undefined,
      "zIndex": undefined,
    }
  }
>
  <TextInput
    editable={true}
    keyboardType="number-pad"
    maxLength={3}
    onChangeText={[Function]}
    placeholder="***"
    style={
      {
        "backgroundColor": undefined,
        "bottom": undefined,
        "color": "rgba(255, 255, 255, 1)",
        "flexGrow": undefined,
        "fontSize": 18,
        "gap": undefined,
        "height": 89,
        "left": undefined,
        "letterSpacing": 0.5,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "maxHeight": undefined,
        "maxWidth": undefined,
        "minHeight": undefined,
        "minWidth": undefined,
        "overflow": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "paddingVertical": undefined,
        "position": undefined,
        "right": undefined,
        "top": undefined,
        "width": "100%",
        "zIndex": undefined,
      }
    }
    testID="input-test-id"
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    alignItemsCenter={true}
    animated={false}
    collapsable={false}
    focusable={true}
    justifyContentCenter={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": undefined,
        "bottom": undefined,
        "flex": 1,
        "flexDirection": "column",
        "flexGrow": undefined,
        "gap": undefined,
        "height": undefined,
        "justifyContent": "center",
        "left": 80,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "maxHeight": undefined,
        "maxWidth": undefined,
        "minHeight": undefined,
        "minWidth": undefined,
        "overflow": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "paddingVertical": undefined,
        "position": "absolute",
        "right": undefined,
        "top": -22,
        "width": undefined,
        "zIndex": undefined,
      }
    }
    testID="info-icon-test-id"
  >
    <View
      style={
        {
          "alignItems": undefined,
          "backgroundColor": undefined,
          "bottom": undefined,
          "flex": undefined,
          "flexDirection": "column",
          "flexGrow": undefined,
          "gap": undefined,
          "height": undefined,
          "justifyContent": undefined,
          "left": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "maxHeight": undefined,
          "maxWidth": undefined,
          "minHeight": undefined,
          "minWidth": undefined,
          "overflow": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "paddingVertical": undefined,
          "position": undefined,
          "right": undefined,
          "top": undefined,
          "width": undefined,
          "zIndex": undefined,
        }
      }
    >
      <SvgMock
        fill="transparent"
        height={46}
        stroke="rgba(34, 63, 70, 1)"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={1}
        width={46}
      />
    </View>
  </View>
</View>
`;
